### ISSUE TITLE
Polymorphic definitions that cannot be instantiated

### ISSUE THREAD

**Issue by SkySkimmer** [coq/coq#9359]
With polymorphic universes it's possible to get a definition which can't be instantiated:
~~~coq
Universes i j.
Polymorphic Definition foo := Type@{i} : Type@{j}.
Constraint j < i.

Fail Check foo.
~~~
The above can also cause coqchk to fail even though coqtop succeeds.

We can also get a couple of definitions with contradictory types which can't be instantiated together (eg going through hurkens), which looks like we've proved false (although we haven't).

Can we find some workaround for this? eg emit a monomorphic instantiation of every polymorphic thing

**Comment by gares** [comment]
I just think that the global names of universes are misleading, they are not global, as the command suggests. If you were forced to bind I and j explicit ly in foo there would be no confusion.

**Comment by SkySkimmer** [comment]
They are global though.

**Comment by JasonGross** [comment]
These universes are monomorphic, right?

Perhaps the monomorpic constraints in polymorphic definitions should be globalised at definition time?

**Comment by JasonGross** [comment]
Yeah, it's not clear to my why there should be local constraints between global universes.  (OTOH, any constraint involving a local universe needs to be local, and there can be constraints between global universes that only show up once you take the transitive closure.  Also, it's nice to be able to easily track where constraints come from.)

**Comment by amintimany** [comment]
I think this is a bug. When you print foo you get the following:
```
Polymorphic foo@{} = Type@{i} : Type@{j}
(*  |= i < j *)
```
This shows that ```i``` and ```j``` are not bound in ```foo``` and therefore the constraint ```i < j``` should be added to the set of global constraints when ```foo``` is processed and no the definition-local set of constraints. @mattam82: do you agree?

**Comment by mattam82** [comment]
I don’t think this is a bug, even if it can be misleading or surprising. There can be constraints mixing local and global universes and this happens quite frequently today as we mix monomorphic definitions from the stdlib with new polymorphic code. Maybe we could warn though if the definition is not polymorphic on anything and suggest to make it monomorphic instead. I suppose coqchk fails because it gathers the whole global constraints of the module/library before checking individual constraints and that’s not faithful to the way coqtop processes definitions in the order of the source file. I don’t see a good fix for that: even extruding all global constraints at each definition would not be enough due to transitivity constraints with some local universes.

**Comment by gares** [comment]
In this case all universes involved in the constraint are global, aren't they?

**Comment by SkySkimmer** [comment]
You can also imagine an example like `Definition foo@{local | i < local, local < j}`

**Comment by mattam82** [comment]
We could, as suggested by Gaëtan, add the transitive closure of the global constraints, that’s a sound approximation I think.

**Comment by ppedrot** [comment]
I also vote for considering this as not a bug. It's definitely surprising and should probably be documented somewhere though.

**Comment by JasonGross** [comment]
@amintimany Given `i < j`, can you instantiate `j` with `i+1`?  If so, for global `u` and `v`, instantiating `v := u+1` would violate `u < i < v`, but would not violate `u < v`.

**Comment by mattam82** [comment]
I was thinking about something along the lines of what Jason says, but it’s
not possible to enforce a specific interval between universes using only
</<= constraints as far as I know, or instantiate with an algebraic like
u+1. You could force u < v or u = v only.
Le lun. 28 janv. 2019 à 19:34, Jason Gross <notifications@github.com> a
écrit :

**Comment by ppedrot** [comment]
It seems that there is now a consensus this is not a bug, let's close.

**Comment by JasonGross** [comment]
Shouldn't we document this before closing?  (Feel free to reclose if not)
